### ISSUE TITLE
Scrap `plugin.yml` commands

### ISSUE THREAD

**Issue by dktapps** [pmmp/PocketMine-MP#6506]
## Description
It's 2024 and most people just declare them in code anyway.

More seriously, command handlers are like event handlers, and the system we create them by should reflect that. It doesn't make much sense to separate the base info from the implementation the way we do right now, and no one likes being bound to `onCommand` (see #5960).

In addition, it's currently impractical to do stuff like using localized descriptions, and `plugin.yml` commands also aren't suited to a future where the command parser actually processes arguments into proper types instead of just passing `string[] $args` to everything.

The only real downside to scrapping these commands is that it won't be possible to tell what commands a plugin has without running the server and doing `/help`. However, since many people already don't use `plugin.yml` for commands, this is kinda a moot point.

We can reduce the friction of creating commands procedurally by implementing changes like #6063.

In the future, we could also implement a system using @doccomments and a magic interface like `Listener` for registering command handlers if there's a demand for it.

## Justification
The whole thing is misconceived. I wrote about this at length in #5960.

This has been discussed for years and years, and I'm frankly amazed that a formal change proposal has never been submitted before.

**Comment by ShockedPlot7560** [comment]
My first thought on this subject is: should we completely remove `onCommand`? As the `ClosureCommand` currently exists, it can be ‘easily’ migrated. If that's the case, the `PluginOwned` interface, the functions derived from it and the PluginOwnedCommand are no longer needed and are destined to disappear with

**Comment by dktapps** [comment]
Yeah, getting rid of `onCommand` is the main motivation for this. `ClosureCommand` provides an easy duct tape fix for plugins that break from this change.

**Comment by dktapps** [comment]
I'm not sure what to do with `PluginOwned` though, I think it may still be of use to have commands signal which plugin they belong to? but in the absence of any requirement to pass a plugin, I'm not sure what to do with it...

**Comment by ShockedPlot7560** [comment]
About `PluginOwned`, as currently, the main use in the core was to register yaml commands, are any developers using it on their side? Because a `CommandExecutor` is needed to work, it will become like `ClosureCommand`: you need a `Closure` instead of a `CommandExecutor`
My opinion would be to remove it too and the `Server::getPluginCommand` etc. will also disappear.
Because linking a command to plugin is optionnal (for now), to be have a right output of the getPluginCommand, we need to have a consistent way to link a command to a plugin, like listeners do. This is probably another part of the command rework

**Comment by dktapps** [comment]
I think at some point I want to be able to have `/help <command>` display information about which plugin it belongs to, and there's also a case to be made for dynamic plugin disabling (open discussion). But I don't see a good way to require a plugin to be provided in the current system.

I'm not sure what the use cases are for getPluginCommand, so I'm not sure if I want to get rid of it or not.

**Comment by ShockedPlot7560** [comment]
Ok, so, PluginOwned have to be changed to suit the deletin of CommandExecutor allowing to keep what depends on PluginOwned.

**Comment by dktapps** [comment]
I don't think it needs to be changed. It's just a decorator. It doesn't matter whether it's used by commands or not.

**Comment by ShockedPlot7560** [comment]
I mean, `PluginOwnedCommand`, sorry

**Comment by dktapps** [comment]
there's no such thing?
